cpp/include/Ice/ServerAuthenticationOptions.h

```diff
@@ -0,0 +1,161 @@
+         * client certificate will be validated using the platform default validation mechanism.
+         *
+         * The validation callback should return true if the certificate chain is valid, false otherwise. Alternatively
+         * an exception can be throw to indicate an error.
+         *
+         * @param context A security context is an opaque data structure that contains security data relevant to the
+         * current connection.
+         * @param info The connection information.
+         * @return true if the certificate is valid, false otherwise.
+         *
+         * [See Manually Validating Schannel
+         * Credentials](https://learn.microsoft.com/en-us/windows/win32/secauthn/manually-validating-schannel-credentials).
+         */
+        std::function<bool(CtxtHandle context, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateValidationCallback;
+#elif defined(__APPLE__)
+        /**
+         * The server's certificate chain.
+         */
+        CFArrayRef serverCertificateChain;
+
+        /**
+         * A callback that allows selecting a certificate chain based on the server's host name. When the callback is
+         * set it has preference over a certificate chain set in serverCertificateChain.
+         *
+         * @param host The server's host name.
+         * @return The server's certificate chain.
+         */
+        std::function<CFArrayRef(const std::string& host)> serverCertificateSelectionCallback;
+
+        /**
+         * The trusted root certificates. If set, the client's certificate chain is validated against these
+         * certificates. If not set the system's root certificates are used.
+         */
+        CFArrayRef trustedRootCertificates;
+
+        /**
+         * A callback that allows manually validating the client certificate chain.
+         *
+         * The validation callback should return true if the certificate chain is valid, false otherwise. Alternatively
+         * an exception can be throw to indicate an error.
+         *
+         * @param trust The trust object that contains the client's certificate chain.
+         * @param info The connection information.
+         * @return true if the certificate chain is valid, false otherwise.
+         */
+        std::function<bool(SecTrustRef trust, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateValidationCallback;
+
+        /**
+         * the requirements for client-side authentication
+         * @see https://developer.apple.com/documentation/security/sslauthenticate
+         */
+        SSLAuthenticate clientCertificateRequired;
+
+        /**
+         * A callback that is called before a new SSL handshake starts. The callback can be used to set additional SSL
+         * context parameters.
+         */
+        std::function<void(SSLContextRef)> sslNewSessionCallback;
```

Comment:
You can also set the certificates from this callback, but the other certificate properties simplify it, and the use case for this callback is for more advanced or less used settings for which we don't provide a property.